app/src/citations.py

```diff
@@ -3,52 +3,111 @@
+    context = get_context(chunks)
+
+    context_list = []
+    for index, chunk_with_subsection in enumerate(context):
+
+        if chunk_with_subsection.chunk.headings:
+            headings_text = "Headings: " + " > ".join(chunk_with_subsection.chunk.headings) + "\n"
+        else:
+            headings_text = ""
+
+        context_text = "Citation: citation-" + str(index) + "\n"
+        context_text += "Document name: " + chunk_with_subsection.chunk.document.name + "\n"
+        context_text += headings_text
+        context_text += "Content: " + chunk_with_subsection.subsection
+
+        context_list.append(context_text)
+
+    return "\n\n".join(context_list)
+
+
+def get_citation_numbers(
+    context: Sequence[ChunkWithSubsection], response: str
+) -> Sequence[ChunkWithSubsection]:
+    """
+    Map (citation-<index>) in `response`, where index is the index in `context`,
+    to a user-friendly citation number.
+    The user friendly citation number is the index of the entry in the returned list, plus one.
+    E.g., if `context` is a list with five entries, and `response` is a string like
+    "Example (citation-3)(citation-1), another example (citation-1).", then this function will return
+    [context[3], [context[1]]; citations referencing context[3] should be shown to the user as "1" and
+    citations referencing context[1] should be shown as "2".
+
+    """
+    citation_indices = re.findall(CITATION_PATTERN, response)
+
+    # Maintain order and avoid duplicates by using a list and tracking seen indices
+    seen = set()
+    citations = []
+
+    for index in citation_indices:
+        idx = int(index)  # Convert string to integer index
+        if 0 <= idx < len(context) and idx not in seen:
+            citations.append(context[idx])
+            seen.add(idx)
+
+    return citations
 
 
 def add_citations(response: str, chunks: list[Chunk]) -> str:
     global _footnote_id
-
     _footnote_id += 1
 
+    context = get_context(chunks)
+    citation_numbers = get_citation_numbers(context, response)
+
     footnote_list = []
 
-    # Replace (chunk-<index>) with the appropriate citation
+    # Replace (citation-<index>) with the appropriate citation
     def replace_citation(match: Match) -> str:
         index = int(match.group(1))
         global _footnote_index
         _footnote_index += 1
         # Leave a citation for chunks that don't exist alone
-        if index >= len(chunks):
-            logger.warning("LLM generated a citation for a chunk ({index}) that doesn't exist.")
-            return f"(chunk-{index})"
+        if index >= len(context):
+            logger.warning("LLM generated a citation for a reference ({index}) that doesn't exist.")
+            return f"(citation-{index})"
+
+        # Find the matching ChunkWithSubsection in the user-friendly
+        # citation numbers and add 1 to get the citation number to display
+        citation_number = 1 + citation_numbers.index(context[index])
```

Comment:
Note to self: this manipulation of the index for user-friendliness gets confusing. In a future refactoring PR, try to separate backend logic from UI user-friendliness.